junifer/pipeline/workdir_manager.py

```diff
@@ -0,0 +1,112 @@
+"""Provide a work directory manager class to be used by pipeline components."""
+
+# Authors: Synchon Mandal <s.mandal@fz-juelich.de>
+#          Federico Raimondo <f.raimondo@fz-juelich.de>
+# License: AGPL
+
+import shutil
+import tempfile
+from pathlib import Path
+from typing import Optional, Union
+
+from ..utils import logger
+from .singleton import singleton
+
+
+@singleton
+class WorkDirManager:
+    """Class for working directory manager.
+
+    This class is a singleton and is used for managing temporary and working
+    directories used across the pipeline by datagrabbers, preprocessors,
+    markers and so on. It maintains a single super-directory and provides
+    directories on-demand and cleans after itself thus keeping the user
+    filesystem clean.
+
+
+    Parameters
+    ----------
+    workdir : str or pathlib.Path, optional
+        The path to the super-directory. If None, "TMPDIR/junifer" is used
+        where TMPDIR is the platform-dependent temporary directory.
+
+    """
+
+    def __init__(self, workdir: Optional[Union[str, Path]] = None) -> None:
+        """Initialize the class."""
+        # Check if workdir is already set
+        if not hasattr(self, "_workdir"):
+            # Convert str to Path
+            if isinstance(workdir, str):
+                workdir = Path(workdir)
+            # Check and set topmost level directory if not provided
+            if workdir is None:
+                workdir = Path(tempfile.gettempdir())
+            logger.debug(f"Setting working directory at {workdir.resolve()!s}")
+            self._workdir = workdir
+
+        # Initialize root temporary directory
+        self._root_temp_dir: Optional[Path] = None
+
+    def __del__(self) -> None:
+        """Destructor."""
+        # Remove root temporary directory
+        if self._root_temp_dir is not None:
+            shutil.rmtree(self._root_temp_dir, ignore_errors=True)
+
+    @property
+    def workdir(self) -> Path:
+        """Get working directory."""
+        return self._workdir
+
+    @workdir.setter
+    def workdir(self, value: Path) -> None:
+        """Set working directory."""
+        logger.debug(f"Changing working directory to {value}")
+        self._workdir = value
```

Comment:
I would not allow to set the workdir like this, for the resasons stated in the constructor